Heap-allocate Font::m_glyphPathMap

smfr opened the pull request:
#### aad1367818261835e2339826089134937752e81d
<pre>
Heap-allocate Font::m_glyphPathMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=251814">https://bugs.webkit.org/show_bug.cgi?id=251814</a>
rdar://105098022

Reviewed by Myles C. Maxfield.

GlyphMetricsMap&lt;std::optional&lt;Path&gt;&gt;&gt; is 896 bytes and rarely used so heap-allocate it, reducing Font from 1256
to 352 bytes.

Also change the return value of Font::pathForGlyph to a Path reference, since this eliminates the need for
the scary comments, and the only caller copies anyway.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::pathForGlyph const):
* Source/WebCore/platform/graphics/Font.h:

Canonical link: <a href="https://commits.webkit.org/259934@main">https://commits.webkit.org/259934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c87506ed3b2fd11ece629547790417a4a9cf2424

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106431 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15459 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115618 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175722 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110338 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6684 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98633 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115286 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12869 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95853 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40435 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94744 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27497 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8691 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28849 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9222 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5418 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14845 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48395 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6864 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10772 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/platform/graphics/Font.h:
```diff
@@ -149,7 +149,7 @@ class Font : public RefCounted<Font> {
     float widthForGlyph(Glyph, SyntheticBoldInclusion = SyntheticBoldInclusion::Incorporate) const;
 
-    const Path& pathForGlyph(Glyph) const; // Don't store the result of this! The hash map is free to rehash at any point, leaving this reference dangling.
+    Path pathForGlyph(Glyph) const;
```
Do we need to do something at the call site to make the caller not perform a second copy?

smfr replied:
I think the return value optimization takes care of this.

smfr committed:
Heap-allocate Font::m_glyphPathMap
https://bugs.webkit.org/show_bug.cgi?id=251814
rdar://105098022

Reviewed by Myles C. Maxfield.

GlyphMetricsMap<std::optional<Path>>> is 896 bytes and rarely used so heap-allocate it, reducing Font from 1256
to 352 bytes.

Also change the return value of Font::pathForGlyph to a Path reference, since this eliminates the need for
the scary comments, and the only caller copies anyway.

* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::pathForGlyph const):
* Source/WebCore/platform/graphics/Font.h:

Canonical link: https://commits.webkit.org/259934@main
